### PR TITLE
Implicit Freezing and `close()` functions

### PR DESCRIPTION
**Pull request by xFrednet**
The number of changes sadly exploded a bit. Most of them are superficial and should hopefully be easy to review.

This PR also changes the stack representation from an `std::vector<>` to actual fields called `_stack[n]` on the frame object. It's not really efficient by any stretch of the imagination, but it allows the invalidation of stack values and works well with our Mermaid output.

### CODE REVIEW

File: src/rt/rt.cc [hunk 1]
```diff
@@ -108,6 +108,7 @@ namespace rt
   objects::DynObject*
   set(objects::DynObject* obj, std::string key, objects::DynObject* value)
   {
+    assert(obj && "Given nullptr, expected a valid DynObject pointer");
```
Comment (mjp41) [inline src/rt/rt.cc:119]: Should this be an actual `error`.  `assert` won't be compiled in Release builds.

File: src/rt/objects/region.h [hunk 2]
```diff
@@ -175,10 +175,8 @@ namespace rt::objects
       collecting = true;
 
       std::cout << "Starting collection" << std::endl;
```
Comment (mjp41) [inline src/rt/objects/region.h:202]: I think using the `for` here is wrong.  The `destruct` call could modify `to_collect`, which will (silently) invalidate the iterator.

File: src/rt/objects/dyn_object.h [hunk 3]
```diff
@@ -323,4 +323,13 @@ namespace rt::objects
     visit(Edge{nullptr, "", start}, pre, post);
   }
 
+  template<typename Pre, typename Post>
+  inline void visit(Region* start, Pre pre, Post post)
+  {
+    for (auto obj : start->objects)
```
Comment (mjp41) [inline src/rt/objects/dyn_object.h:356]: Could `visit` modify `start->objects`?

**Comment by xFrednet** [comment]
Good comments, shows that I'm still not used to C++ and why memory safe languages are a good idea... Anyhow, let's get this merged :D

**Pull request merged**